WIP: Feature/pkcs11 token

MPeisl opened the pull request:
Support for generic PKCS#11 tokens:
- scd/pkcs11-lib/*: contains headers and helper libraries
- scd/pkcs11-test/*: unit-tests for p11token module
- scd/p11token: actual scd_token implementation for PKCS#11-API

~~TODO: integration in daemon~~
TODO: integration testing
TODO before merging: squash commits

quitschbo commented on daemon/c_smartcard.c:
```diff
@@ -1075,13 +1083,22 @@ c_smartcard_new(compartment_t *compartment)
+	if (smartcard->token_type == CONTAINER_TOKEN_TYPE_PKCS11) {
+		if (NULL == container_get_pkcs11_module(smartcard->container)) {
+			ERROR("PKCS#11 module missing in container config. Abort creation of container");
+			mem_free0(smartcard);
```
return NULL;

MPeisl replied:
Done

quitschbo commented on daemon/c_smartcard.c:
```diff
@@ -1075,13 +1083,22 @@ c_smartcard_new(compartment_t *compartment)
 	smartcard->sock = sock_unix_create_and_connect(SOCK_SEQPACKET, scd_sock_path);
 	if (smartcard->sock < 0) {
-		mem_free0(smartcard);
+		mem_free0(smartcard); //? potential mem_leak for usbtoken
```
yes smartcard->token_serial may be allocated here. Thus should be freed, same counts for smartcard->pkcs11_module. May instead of return NULL in error cases. We introduce a jump mark err:
and do cleanup there

MPeisl replied:
Done

quitschbo commented on daemon/container.h:
```diff
@@ -377,6 +378,9 @@ container_get_dev_assign_list(const container_t *container);
 container_type_t
 container_get_type(const container_t *container);
 
+const char *
```
add a short comment about what function does

quitschbo commented on scd/p11token.h:
```diff
@@ -0,0 +1,131 @@
+#include "pkcs11-lib/pkcs11.h"
+#include <stddef.h>
+
+// todo: move to c-file -> hide implementation details to user
```
remove comment

quitschbo commented on scd/p11token.h:
```diff
@@ -0,0 +1,131 @@
+#define P11TOKEN_H
+
+#include <stdbool.h>
+#include "pkcs11-lib/pkcs11.h"
```
I think this header is not used here, -> move this to p11token.h

MPeisl replied:
Correct, header was only needed for struct p11token which is now defined in p11token.c. Will be moved.

quitschbo commented on scd/scd.c:
```diff
@@ -481,13 +483,17 @@ scd_token_new(const DaemonToToken *msg)
 
 	create_data.type = scd_proto_to_tokentype(msg);
 
+	// ? why no switch statement here
```
yes a "switch" would look nicer here for me, too :)

quitschbo commented on scd/token.c:
```diff
@@ -585,6 +659,50 @@ token_new(const token_constr_data_t *constr_data)
+		new_token->reset_auth = int_reset_auth_p11;
+		new_token->get_atr = int_get_atr_p11;
+		new_token->send_apdu = int_send_apdu_p11;
+		break;
```
For a followup refactoring, we should consider to introduce a struct for all those functions and hide
the token specific functions in the corresponding p11token.c/usbtoken.c files (analogues to the compartment_module_t)

quitschbo commented on scd/pkcs11-lib/libpkcs11.c:
```diff
@@ -0,0 +1,125 @@
+#include <stdio.h>
+#include <string.h>
+
+#include "libscdl.h"
```
get rid of this wrapper lib, we do not need win32 compat.
directly call dlopen, dlsym and dlclose in the following.

MPeisl replied:
Done

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,623 @@
+	return -1;
+}
+
+// internal helper functions
```
put those functions at the beginning of the c file where you now have the prototype definitions.
remove the prototype definitions and declare those functions static.

quitschbo commented on scd/Makefile:
```diff
@@ -92,6 +93,14 @@ ifeq ($(SYSTEMD),y)
 		-DSCD_CONTROL_SOCKET=NULL
 endif
+ifeq ($(PKCS11), y)
+    # If requested, we build sc-hsm support into trustme
```
\# If requested, we build pkcs11 support into the scd

MPeisl committed:
scd/token,softtoken,usbtoken: unify usage of 'const'

This commit unifies the usage of the 'const' qualifier in the token API.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
scd/token,softtoken,usbtoken: unify usage of 'const'

This commit unifies the usage of the 'const' qualifier in the token API.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
scd/*: PKCS#11 Support

The scd now supports a new tokentype which is
build upon PKCS#11. In order to use the new token type a path to
a valid PKCS#11-Module needs to be specified In order to use the new
token type a path to
a valid PKCS#11-Module needs to be specified.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
scd/*: PKCS#11 Support

Add a new PKCS11-backed token (TOKEN_TYPE_PKCS11) implementation to scd.
Support is enabled at compile time using the PKCS11 build flag.
PKCS#11-modules are loaded dynamically at runtime based on a
configurable module path.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
daemon/*: Support new PKCS#11 tokentype

Add support for new tokentype offered by scd. Add new 'pkcs11_module'
configuration option. Similar to usbtoken or softtoken one can now
configure a container to use a specific PKCS#11 module.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
daemon/*: Support new PKCS#11 tokentype

Add support for new tokentype offered by scd. Add new 'pkcs11_module'
configuration option. Similar to usbtoken or softtoken one can now
configure a container to use a specific PKCS#11 module.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
daemon/*: Make scd-unit env configurable

Pass scd-specific environment variables defined in 'device.conf' to scd-unit.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
daemon/*: Make scd-unit env configurable

Pass scd-specific environment variables defined in 'device.conf' to scd-unit.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

quitschbo commented on daemon/device_config.h:
```diff
@@ -91,4 +91,10 @@ device_config_get_audit_size(const device_config_t *config);
 bool
 device_config_get_tpm_enabled(const device_config_t *config);
+
+char **
```
make this const

quitschbo commented on daemon/scd.c:
```diff
@@ -202,7 +203,7 @@ scd_tpm2d_inotify_written_cb(const char *path, uint32_t mask, event_inotify_t *i
 
 int
-scd_init(void)
+scd_init(char **env, size_t env_len)
```
const char

MPeisl committed:
daemon/*: make init_env const

Refactor init_env parameters to const char** in container and compartment
functions.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl commented:
Refactored module loading, addressed change request. This PR currently depends on #586

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+	// An open session handle is the ground truth for "already unlocked".
+	// Do not rely on the generic token->locked flag here: p11token_unlock may
+	// be invoked directly (e.g. from p11token_change_pin) without going through
+	// the token_unlock() wrapper that maintains that flag.
```
multi-line comments pleas use following style:
```
/*
 * line1
 * line2
 */
```

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+	// flag: when called directly (e.g. from p11token_change_pin) that flag is
+	// not maintained, and keying off it would skip C_Logout/C_CloseSession/
+	// C_Finalize/dlclose, leaking the session and leaving the module
+	// C_Initialize'd (breaking the next unlock with CKR_CRYPTOKI_ALREADY_INITIALIZED).
```
multi-line comment

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+} p11_cipher_params_t;
+
+int
+int_pkcs11_wrap_to_pkcs7(const unsigned char *ciphertext, size_t ct_len,
```
declare static, isn't exposed by header

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+}
+
+int
+int_pkcs7_unwrap_to_pkcs11(const unsigned char *der, int der_len, p11_cipher_params_t *params,
```
declare static, isn't exposed by header

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+
+// TODO: use pairing secret?
+token_err_t
+p11token_unlock(void *int_token, const char *passwd, UNUSED const unsigned char *pairing_secret,
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+}
+
+token_err_t
+p11token_lock(void *int_token)
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+
+token_err_t
+p11token_wrap_key(void *int_token, UNUSED const char *label, unsigned char *plain_key,
+		  size_t plain_key_len, unsigned char **wrapped_key, int *wrapped_key_len)
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+}
+
+token_err_t
+p11token_unwrap_key(void *int_token, UNUSED const char *label, unsigned char *wrapped_key,
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+}
+
+token_err_t
+p11token_change_pin(void *int_token, const char *oldpass, const char *newpass,
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+}
+
+tokentype_t
+p11token_get_type()
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+ * lock token and cleanup data structure
+*/
+void
+p11token_free(void *int_token)
```
static

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+error:
+	if (pt) {
+		mem_free0(pt);
+	}
```
`if (ct) mem_free0(ct)`; evtl. mem_memset0 beforehand (is encrypted anyway but wrapped key)

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+	IF_TRUE_GOTO_ERROR(pt_len > INT_MAX, error);
+	*plain_key = pt;
+	*plain_key_len = (int)pt_len;
+
```
`if (ct) mem_free0(ct)`; evtl. mem_memset0 beforehand (is encrypted anyway but wrapped key)

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+	P11_CHECK_RV_GOTO(ctx->C_CloseSession(sh), error);
+
+	// free slot id ptr
+	mem_free(slot);
```
possible double free if next check failes and goes to error_init.
move mem_free(slot) after the Unload Module block? Double safe with mem_free0

quitschbo commented on scd/p11token.c:
```diff
@@ -0,0 +1,1108 @@
+	token->ctx = NULL;
+	token->sh = NULL;
+	token->module = NULL;
+	token->initialized = true;
```
mem_free0(slot)

MPeisl committed:
daemon/*_config: Guard pkcs11-changes for CC_MODE

Guard PKCS#11 related changes to the protobuf files such that they are
excluded in CC_MODE builds.

Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
daemon/*,scd/*: Improvements to PKCS#11 support

This commit contains several enhancements and security hardening
improvements to the PKCS#11 token type.
Parts are based on severall rounds of AI-based security audits
with corresponding code suggestions.
All proposed changes have been manually reviewed and applied on a
per case basis.

Assisted-by: Claude:claude-fable-5
Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
scd/p11token: add static to all internal functions

MPeisl committed:
address code review: fix mem-errors

MPeisl committed:
common/ssl_util: Add pbkdf2 helper

Add key derivation function which implements
PBKDF2-HMAC-SHA256.

Assisted-by: Claude:claude-fable-5
Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>

MPeisl committed:
scd/p11token: Bind pin to device secret

Bind user-pin to pairing device secret using a KDF.

Assisted-by: Claude:claude-fable-5
Signed-off-by: Maximilian Peisl <maximilian.peisl@aisec.fraunhofer.de>
